examples/tensorflow_distributed/tf_distributed.yaml

```diff
@@ -0,0 +1,53 @@
+# Distributed TensorFlow training on GCP with 2 V100 GPUs
+#
+# Uses MultiWorkerMirroredStrategy to distribute training across multiple nodes with SkyPilot.
+#
+# Usage:
+#   sky launch -c myclus tf_distributed.yaml
+#   sky down myclus
+
+resources:
+  cloud: gcp
+  accelerators: V100:1 # Provision 1 V100 GPU per node
+
+# Provision 2 nodes, giving us a total of 2 GPUs in the cluster
+num_nodes: 2
+
+# Copy files (train.py) from the current working directory to all nodes on the cluster
+workdir: .
+
+# Install dependencies on all nodes during the setup phase
+setup:
+  pip install tensorflow==2.11.0
+
+# The following shell commands are run on all nodes at execution
+run: |
+  # Port to use for tensorflow. Can be any unused port.
+  PORT=2222
+  
+  # ======== Construct TF_CONFIG ========
+  # Since we are using MultiWorkerMirroredStrategy, we need to construct a TF_CONFIG
+  # environment variable that contains the list of all worker IPs and the current node's rank.
+  # SkyPilot provides the SKYPILOT_NODE_IPS and SKYPILOT_NODE_RANK environment variables to get this information.
+  # 
+  # Examples of envvars:
+  #   SKYPILOT_NODE_IPS="192.168.0.1 192.168.0.2"
+  #   SKYPILOT_NODE_RANK="1"
+  
+  IP_LIST_IDX=$(($SKYPILOT_NODE_RANK + 1))
+  echo "$SKYPILOT_NODE_IPS" | sed -n ${IP_LIST_IDX}p
+  
+  # Construct TF_CONFIG worker list
+  WORKER_LIST=""
+  for IP in $(echo "$SKYPILOT_NODE_IPS" | sed "s/,/ /g")
+  do
+    WORKER_LIST="$WORKER_LIST\"$IP:$PORT\","
+  done
+  # Remove trailing comma
+  WORKER_LIST=$(echo $WORKER_LIST | sed 's/,$//')
+  
+  export TF_CONFIG="{\"cluster\": {\"worker\": [$WORKER_LIST]}, \"task\": {\"type\": \"worker\", \"index\": $SKYPILOT_NODE_RANK}}"
+  echo $TF_CONFIG
+  
```

Comment:
Nice! Glad to see this can be done in bash. I am wondering if it is easier to understand for ML people if we do it in python like the following:

```
    python -u <<- EOF
        import json
        import os
        
        port = 2222
        node_ips_str = os.environ.get(SKYPILOT_NODE_IPS)
        node_ips = node_ips_str.strip().split('\n')

        node_rank = os.environ.get(SKYPILOT_NODE_RANK)
        worker_list = [f'{ip}:{port}' for ip in node_ips] 
        tf_config = {
                'cluster': {
                    'worker': worker_list,
                },
                'task': {
                    'type': 'worker',
                    'index': node_rank
                }
            }
        with open('/tmp/{os.environ.get("SKYPILOT_JOB_ID"', 'w') as f:
            json.dump(tf_config, f)
    EOF
    export TF_CONFIG=$(cat /tmp/$SKYPILOT_JOB_ID)
    echo $TF_CONFIG
```

Reply:
That's a great idea. Python is definitely easier to understand. I've switched to python in the YAML.